enterprise/tailnet/pgcoord.go

```diff
@@ -520,7 +518,14 @@ func (b *binder) handleBindings() {
 	for {
 		select {
 		case <-b.ctx.Done():
-			b.logger.Debug(b.ctx, "binder exiting", slog.Error(b.ctx.Err()))
+			b.logger.Debug(b.ctx, "binder exiting, updating peers to lost", slog.Error(b.ctx.Err()))
+			err := b.store.UpdateTailnetPeerStatusByCoordinator(context.Background(), database.UpdateTailnetPeerStatusByCoordinatorParams{
```

Comment:
I picked the binder as the place to perform this update since it's responsible for other types of updates to peers but I could see it making more since being closer to the actual shutdown of the coordinator.

Reply:
I think the binder is the right place to perform this update, but I'm pretty sure there is a race condition between this query and the rest of the binding worker goroutines.  You set everything lost here, but then a worker in flight updates one of the peers so it's no longer lost.

This should be the very last query the binder sends, and so it needs to wait for the worker wait group to finish.

Reply:
I figured the `b.ctx` would save me here but there may be a small window where what you talked about can happen and the wait group definitely makes it more obvious in any case. I think there's another issue here where we race with the shutdown of the process. Since this isn't coupled with the `wg` there's nothing preventing `pgcoord.Close` from returning prior to use updating all peers right?

Reply:
Prior to your changes:

1. `pgcoord.Close()` cancels the context for the binder, tunneler, handshaker, and connIOs. This causes them to start tearing down.
2. When binder, tunneler, and handshaker workers have all shut down, this triggers us to cancel the context for the querier & heartbeats.
3. `pgcoord.Close()` waits for querier and heartbeats to shut down.

With these changes:

I don't think we need to have a staggered shut down of the different components: binder, tunneler, handshaker, querier, heartbeats, and connIOs can all shut down in parallel.

_Within_ the binder, we should wait for the workers to stop before setting everything lost.

We should have `pgcoord.Close()` wait for all the goroutines to stop before returning.  Eventual consistency means that we can tolerate leaving the DB in a bad state, but I think it's better to make an effort to get things in the right state, and to wait for goroutines to finish: if nothing else it helps prevent test flakes & `gorace` errors.